File: java/450.delete-node-in-a-bst.java
/*
 * @lc app=leetcode id=450 lang=java
 *
 * [450] Delete Node in a BST
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode deleteNode(TreeNode root, int key) {
        if (root == null) return null;
        if (root.val > key) root.left = deleteNode(root.left, key);
        else if (root.val < key) root.right = deleteNode(root.right, key);
        else {
            if (root.left == null || root.right == null) {
                root = (root.left == null) ? root.right : root.left;
            } else {
                TreeNode curNode = root.right;
                while (curNode.left != null) curNode = curNode.left;
                root.val = curNode.val;
                root.right = deleteNode(root.right, curNode.val);
            }
        }
        
        return root;
    }
}

File: go/450.delete-node-in-a-bst.go
/*
 * @lc app=leetcode id=450 lang=golang
 *
 * [450] Delete Node in a BST
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func deleteNode(root *TreeNode, key int) *TreeNode {
	if root == nil {
		return root
	}
	
	if key < root.Val {
		root.Left = deleteNode(root.Left, key)
	} else if key > root.Val {
		root.Right = deleteNode(root.Right, key)
	} else {

		if root.Right == nil {
			return root.Left
		} else if root.Left == nil {
			return root.Right
		} else {

			prev, tmp := root, root.Right
			if tmp.Left == nil {
				root.Val = tmp.Val
				root.Right = tmp.Right
			} else {
				for tmp.Left != nil {
					prev = tmp
					tmp = tmp.Left
				}
				root.Val = tmp.Val
				prev.Left = tmp.Right
			}
		}
	}
	
	return root
}
// @lc code=end

